### ISSUE TITLE
PRIMED key grows indefinitely when job re-queues itself in `after_unlock` callback

### ISSUE THREAD

**Issue by ttstarck** [mhenrixon/sidekiq-unique-jobs#787]
**Describe the bug**
`PRIMED` key grows indefinitely  for jobs that requeue themselves in `after_unlock` callback if the job is completed and re-queued again within 1 second of when the initial job was queued.

**Expected behavior**
`PRIMED` key is removed or stays at a small length when job is fully locked.

**Current behavior**
`PRIMED` key grows indefinitely, with `"1"` as a value when job is locked.

**Worker class**
```ruby
class AfterUnlockWorker
  include Sidekiq::Job
  sidekiq_options lock: :until_executed

  def perform(args)
    @args = args
  end

  def after_unlock
    self.class.perform_async(@args)
  end
end
```

**Additional context**
Sidekiq Version: 6.5.8
SidekiqUniqueJobs Version: 7.1.29

#### Rspec Example:
```ruby
# frozen_string_literal: true

require "rspec"
require "sidekiq"
require "sidekiq-unique-jobs"
require "sidekiq_unique_jobs/testing"

Sidekiq.configure_client do |config|
  config.client_middleware do |chain|
    chain.add(SidekiqUniqueJobs::Middleware::Client)
  end
end

RSpec.configure do |config|
  config.before(:each) do
    Sidekiq::Worker.clear_all
    Sidekiq::Testing.server_middleware do |chain|
      chain.add(SidekiqUniqueJobs::Middleware::Server)
    end
  end
end

RSpec.describe "PRIMED key spec" do
  it "grows PRIMED key indefinitely" do
    Sidekiq::Testing.fake!
    AfterUnlockWorker.perform_async(1)
    lock_digest = AfterUnlockWorker.jobs.first["lock_digest"]
    primed_key_length = -> { Sidekiq.redis { |r| r.llen("#{lock_digest}:PRIMED") } }
    primed_key_values = -> { Sidekiq.redis { |r| r.lrange("#{lock_digest}:PRIMED", 0, -1) }.uniq }
    100.times do |i|
      AfterUnlockWorker.perform_one
      puts "Test #{i} - Length of PRIMED key #{primed_key_length.call}" if i % 10 == 0
    end
    puts "Test Finished - Length of PRIMED key #{primed_key_length.call}"
    puts "Test Finished - Unique PRIMED key values #{primed_key_values.call}"
    sleep(1)
    puts "Test Finished - Length of PRIMED key after 1 second sleep #{primed_key_length.call}"
  end
end
```

Output from running the above spec:
```
Test 0 - Length of PRIMED key 1
Test 10 - Length of PRIMED key 11
Test 20 - Length of PRIMED key 21
Test 30 - Length of PRIMED key 31
Test 40 - Length of PRIMED key 41
Test 50 - Length of PRIMED key 51
Test 60 - Length of PRIMED key 61
Test 70 - Length of PRIMED key 71
Test 80 - Length of PRIMED key 81
Test 90 - Length of PRIMED key 91
Test Finished - Length of PRIMED key 100
Test Finished - Unique PRIMED key values ["1"]
Test Finished - Length of PRIMED key after 1 second sleep 0
```

This looks to be because when we run the `unlock` LUA script on a job, we push `"1"` to the `QUEUED` key: 
https://github.com/mhenrixon/sidekiq-unique-jobs/blob/main/lib/sidekiq_unique_jobs/lua/unlock.lua#L95-L96

I unfortunately don't have enough context around the locking/unlocking algorithm to know if this is actually necessary or not.

I will say that after doing some more reading, [Redis suggests a much simpler algorithm](https://redis.io/docs/manual/patterns/distributed-locks/#correct-implementation-with-a-single-instance) for setting up Distributed locking for a single master redis instance (which I think most people are running).

I'd love to hear about the reasoning behind why the `QUEUED` and `PRIMED` keys were added.

**Comment by mhenrixon** [comment]
Yes, redis suggests a simpler algorithm, but I needed more, so I dreamed up my own algorithm. 

People requested that they lock only during runtime, not during queuing, meaning sidekiq picks the job up, runs it through the server middleware, and the server constrains any duplicates from running simultaneously.

People also requested that the locks stay active as long as the job is in Sidekiq. If you want something simpler, I recommend https://github.com/rwojsznis/sidekiq-lock or sidekiq-pro if you have the money. 

I don't think you should schedule a new job as part of the old one still running. Sidekiq needs to be done processing the current job when you push the next one. Preferably, you should do that outside of this "Redis transaction" or "unit of work."

The queued and primed keys are used to block as part of processing the job. The push of "1" is accompanied by an expiration to ensure that listeners on the queued key can keep listening. 

There are things I can do to improve your situation, but you know you are doing an infinite recursion, right?

**Comment by gsmetal** [comment]
I don't know if it's the same problem or not, but we also have the PRIMED keys continuous grow. 

We don't use rescheduling in after_unlock and mostly use `lock: :until_executed, on_conflict: :log` for our jobs, and we have continuously growing number of non-expiring keys in Redis. I looked into them and there are only `:PRIMED` keys there:

```
# redis-cli -n 1 --scan | while read LINE ; do TTL=`redis-cli -n 1 ttl "$LINE"`; if [ $TTL -eq  -1 ]; then echo "$LINE"; fi; done;
uniquejobs:1a8fde3b4a48e2bc57b5c932d08383e2:PRIMED
uniquejobs:8f20c9b895f31114b8ba5b78e236dd90:PRIMED
uniquejobs:64d7ad51c9792e818e8112fa160d5982:PRIMED
uniquejobs:ee606fbd589b94cc28e0d4bedccc13be:PRIMED
uniquejobs:4244c6a2c359262959693083109f5f2d:PRIMED
uniquejobs:69a05fbf85237ce7930883a621f0c387:PRIMED
uniquejobs:7ae62fe173ffdd6919d3bbb7ebb8d6f3:PRIMED
uniquejobs:9a53755e27b128399c7783717d98c1a7:PRIMED
uniquejobs:5a9f36166194426f87add13492bedd84:PRIMED
uniquejobs:db5bf5aa573a0eee170f94d7340c2f8d:PRIMED
uniquejobs:1756de4b515e31f1928f2ee5878d8bd5:PRIMED
uniquejobs:bdb52a8f47d797cff94e8bee28610280:PRIMED
uniquejobs:e7e2d275c352aa86e727a9fe37c3b1a7:PRIMED
...
```

I've tried to figure out where the problem could come from, but with no luck.

**Comment by mhenrixon** [comment]
Yeah sorry about that @gsmetal, I still haven't been able to consistently reproduce this problem. It seems some people, for whaever reason, runs into this problem. 

I'll keep searching!

**Comment by jamesrusso** [comment]
Cross-referencing #998, which is a different route into what I think is the same weakness. Your diagnosis here looks right to me for this case.

The common part is that `lock.lua` only ever removes `job_id` from `primed`, so any token that is not this client's own jid survives that `LREM`. The `"1"` sentinel `unlock.lua` pushes onto `:QUEUED` is exactly such a token, and a client that pops it instead of its own jid leaves it behind. The only thing that ever clears it is the blanket `PEXPIRE primed 1000` on the success path, which is why you see the list grow when a job requeues itself inside that one second.

#998 is the case where that `PEXPIRE` never runs at all: the `limit_exceeded` branch returns before both the `LREM` and the `PEXPIRE`, so the entry gets no expiry and is permanent rather than merely outpaced.

Worth noting for anyone landing here: the patch I proposed on #998 only fixes the refusal branch, so it would not help this report. Something narrower, like removing the popped token whatever it was rather than only `job_id`, would cover both. Still present in 8.1.0; the v9 rewrite drops the queued/primed handshake entirely.
